src/planner/space_constraint.c

```diff
@@ -0,0 +1,283 @@
+}
+
+static OpExpr *
+transform_space_constraint(PlannerInfo *root, List *rtable, OpExpr *op)
+{
+	Var *var = linitial_node(Var, op->args);
+	Const *value = lsecond_node(Const, op->args);
+	Const *part_value;
+	RangeTblEntry *rte = list_nth(rtable, var->varno - 1);
+	Dimension *dim = get_space_dimension(rte->relid, var->varattno);
+	PartitioningFunc partfunc = dim->partitioning->partfunc;
+	TypeCacheEntry *tce = lookup_type_cache(partfunc.rettype, TYPECACHE_EQ_OPR);
+
+	/* build FuncExpr to use in eval_const_expressions */
+	FuncExpr *partcall = make_partfunc_call(partfunc, list_make1(value), var->varcollid);
+
+	/*
+	 * We should always be able to constify here
+	 */
+	part_value = castNode(Const, eval_const_expressions(root, (Node *) partcall));
+
+	/* build FuncExpr with column reference to use in constraint */
+	partcall->args = list_make1(copyObject(var));
+
+	OpExpr *ret = (OpExpr *) make_opclause(tce->eq_opr /* opno */,
+										   BOOLOID /*opresulttype */,
+										   false /* opretset */,
+										   (Expr *) partcall /* left */,
+										   (Expr *) part_value /* right */,
+										   InvalidOid /* opcollid */,
+										   InvalidOid /* inputcollid */);
+	ret->location = -29811;
+	return ret;
+}
+
+static ScalarArrayOpExpr *
+transform_scalar_space_constraint(PlannerInfo *root, List *rtable, ScalarArrayOpExpr *op)
+{
+	Var *var = linitial_node(Var, op->args);
+	RangeTblEntry *rte = list_nth(rtable, var->varno - 1);
+	Dimension *dim = get_space_dimension(rte->relid, var->varattno);
+	PartitioningFunc partfunc = dim->partitioning->partfunc;
+	TypeCacheEntry *tce = lookup_type_cache(partfunc.rettype, TYPECACHE_EQ_OPR);
+	List *part_values = NIL;
+	ListCell *lc;
+
+	/* build FuncExpr to use in eval_const_expressions */
+	FuncExpr *partcall = make_partfunc_call(partfunc, NIL, var->varcollid);
+
+	foreach (lc, lsecond_node(ArrayExpr, op->args)->elements)
+	{
+		Const *value = lfirst_node(Const, lc);
+		/*
+		 * We can skip NULL here as elements are ORed and partitioning dimensions
+		 * have NOT NULL constraint.
+		 */
+		if (!value->constisnull)
+		{
+			List *args = list_make1(value);
+			partcall->args = args;
+			part_values = lappend(part_values,
+								  castNode(Const, eval_const_expressions(root, (Node *) partcall)));
+		}
+	}
+	/* build FuncExpr with column reference to use in constraint */
+	partcall->args = list_make1(copyObject(var));
+
+	ArrayExpr *arr2 = makeNode(ArrayExpr);
+	arr2->array_collid = InvalidOid;
+	arr2->array_typeid = get_array_type(partfunc.rettype);
+	arr2->element_typeid = partfunc.rettype;
+	arr2->multidims = false;
+	arr2->location = -1;
+	arr2->elements = part_values;
+
+	ScalarArrayOpExpr *op2 = makeNode(ScalarArrayOpExpr);
+	op2->opno = tce->eq_opr;
+	op2->args = list_make2(partcall, arr2);
+	op2->inputcollid = InvalidOid;
+	op2->useOr = true;
+	op2->location = PLANNER_LOCATION_MAGIC;
+
+	return op2;
+}
+
+Node *
+ts_amend_space_constraints(PlannerInfo *root, List *rtable, Node *node)
```

Comment:
Let's add a top-level description of the algorithm here. If we have an equality condition on a space partitioning column, we also add a corresponding condition on `get_partition_hash` on this column. These conditions match the constraints on chunks, so Postgres' constraint exclusion is able to use them and exclude the chunks. Did I get it right?

`amend` sounds slightly imprecise, makes me think that we modify something, but here we just add a different form that matches the constraints on chunks. Maybe `ts_add_space_constraints`? Anyway, not so important.

Reply:
TBH I would expect the Postgres planner to be smart enough to prove that, if you have an immutable function in a constraint and a constant condition in WHERE...

Reply:
`operator_predicate_proof` doesn't evaluate sides of the constraint and only matches costraint & clause verbatim. And `eval_const_expression` we do on the relation constraints can't use equivalence classes to evaluate Vars to const. So no, it's not smart enough :(

Reply:
done